### DISCUSSION TITLE
Change background color for a row in Datagrid

### DISCUSSION THREAD

**Discussion by ShrutiJaiswal1494** [AvaloniaUI/Avalonia#5810] [Q&A]
Hi,

I have a Datagrid which is bound to an ObservableCollection. Now, based on the text that is entered in a textbox I want to highlight the rows that matches it. I am not able to find a way to highlight a particular row in DataGrid. Is there a way to do that?

**Comment by FoggyFinder** [comment]
If you mean highlightning aside of selection I think it should be feasible with converter. I can't write an example at the moment but I'll try to edit my comment later.

**Reply by ShrutiJaiswal1494** [reply]
Thanks for the reply. Yeah I want to highlight rows other than the selected ones. If you could share an example that would be great.

Is it possible to extract rows of DataGrid as DataGridRows and then use it to edit the background. I want to do something similar like this.

```
foreach (var item in dataGrid1.Items)
{
      var row = dataGrid1.ItemContainerGenerator.ContainerFromItem(item) as DataGridRow;
      row.Background = Brushes.Transparent;
}
```

**Reply by maxkatz6** [reply]
> Is it possible to extract rows of DataGrid as DataGridRows

No, and shouldn't be. If you have 1000 items, DataGrid will create only minimal number of rows to display them and recycle. 
You might need LoadingRow event, which is called when specific row is attached to new item. You can define bindings there, but do not forget to dispose them, when UnloadingRow is invoked.

**Reply by maxkatz6** [reply]
Also I haven't tested, but some style like that might work:
```
<Style Selector="DataGrid#MyDataGrid DataGridRow">
    <Setter Property="Background" Value="{Binding RowItemProp, Converter={StaticResource ValueToColorConverter}}" />
</Style>
```

**Reply by FoggyFinder** [reply]
Seems like direct converter is not an option. For some reason background's change doesn't refresh DataGrid. So probably adding a property to VM side is a way to go.

Though I could miss something so below a simple sample that doesn't work despite converter returns correct value:

Style
```xml
    <Style Selector="DataGrid DataGridRow">
      <Setter Property="Background">
        <Setter.Value>
          <MultiBinding Converter="{StaticResource ResourceKey=p2BConv}">
            <MultiBinding.Bindings>
              <Binding Path="."/>
              <Binding Path="Text" ElementName="inputTxb"/>
            </MultiBinding.Bindings>
          </MultiBinding>
        </Setter.Value>
      </Setter>
    </Style>
```

DataGrid

```xml
    <DataGrid Grid.Row="1" Items="{Binding Data}" >
      <DataGrid.Columns>
        <DataGridTextColumn
            Width="*"
            Binding="{Binding Value}"
            Header="Value" />
        <DataGridTextColumn
            Width="*"
            Binding="{Binding Text}"
            Header="Text" />
      </DataGrid.Columns>
    </DataGrid>
```
Item
```cs
    public class SomeType
    {
        public SomeType(int value, string text)
        {
            Value = value;
            Text = text;
        }

        public int Value { get; }

        public string Text { get; }
    }
```

Converter
```cs
    public class PropertyToBrushConverter : IMultiValueConverter
    {
        public object Convert(IList<object> values, Type targetType, object parameter, CultureInfo culture)
        {
            if (values[0] is SomeType st && values[1] is string str)
            {
                if (st.Text.Contains(str))
                {
                    return Brushes.LightYellow;
                }
                return Brushes.LightGray;
            }
            return Brushes.LightGray;
        }

        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
        {
            throw new NotImplementedException();
        }
    }
```
